crates/matrix-sdk/src/room/timeline/pagination.rs

```diff
@@ -0,0 +1,140 @@
+use std::fmt;
+
+/// Options for pagination.
+pub struct PaginationOptions<'a> {
+    inner: PaginationOptionsInner<'a>,
+}
+
+impl<'a> PaginationOptions<'a> {
+    /// Do a single pagination request, asking the server for the given
+    /// maximum number of events.
+    ///
+    /// The server may choose to return fewer events, even if the start or end
+    /// of the visible timeline is not yet reached.
+    pub fn single_request(event_limit: u16) -> Self {
+        Self::new(PaginationOptionsInner::SingleRequest { event_limit_if_first: Some(event_limit) })
+    }
+
+    /// Continually paginate with a fixed `limit` until at least the given
+    /// amount of timeline items have been added (unless the start or end of the
+    /// visible timeline is reached).
+    ///
+    /// The `event_limit` represents the maximum number of events the server
+    /// should return in one batch. It may choose to return fewer events per
+    /// response.
+    pub fn until_num_items(event_limit: u16, items: u16) -> Self {
+        Self::new(PaginationOptionsInner::UntilNumItems { event_limit, items })
+    }
+
+    /// Paginate once with the given initial maximum number of events, then
+    /// do more requests based on the user-provided strategy
+    /// callback.
+    ///
+    /// The callback is given numbers on the events and resulting timeline
+    /// items for the last request as well as summed over all
+    /// requests in a `paginate_backwards` call, and can decide
+    /// whether to do another request (by returning
+    /// `Some(next_event_limit)`) or not (by returning `None`).
+    pub fn custom(
+        initial_event_limit: u16,
+        pagination_strategy: impl FnMut(PaginationOutcome) -> Option<u16> + 'a,
+    ) -> Self {
+        Self::new(PaginationOptionsInner::Custom {
+            event_limit_if_first: Some(initial_event_limit),
+            strategy: Box::new(pagination_strategy),
+        })
+    }
+
+    pub(super) fn next_event_limit(
```

Comment:
Note to self: for `SingleRequest` and `Custom`, this should use `Option::take`. Probably related to the test failure.